sydent/hs_federation/verifier.py

```diff
@@ -0,0 +1,206 @@
+        }
+
+    @defer.inlineCallbacks
+    def _getEndpointForServer(self, server_name):
+        if ':' in server_name:
+            defer.returnValue(tuple(server_name.rsplit(':', 1)))
+
+        service_name = "%s.%s.%s" % ('_matrix', '_tcp', server_name)
+
+        default = server_name, 8448
+
+        try:
+            answers, _, _ = yield twisted.names.client.lookupService(service_name)
+        except DNSNameError:
+            logger.info("DNSNameError doing SRV lookup for %s - using default", service_name)
+            defer.returnValue(default)
+
+        for answer in answers:
+            if answer.type != twisted.names.dns.SRV or not answer.payload:
+                continue
+
+            # XXX we just use the first
+            logger.info("Got SRV answer: %r / %d for %s", str(answer.payload.target), answer.payload.port, service_name)
+            defer.returnValue((str(answer.payload.target), answer.payload.port))
+
+        logger.info("No valid answers found in response from %s (%r)", server_name, answers)
+        defer.returnValue(default)
+
+    @defer.inlineCallbacks
+    def _getKeysForServer(self, server_name):
+        """Get the signing key data from a home server.
+        """
+
+        if server_name in self.cache:
+            cached = self.cache[server_name]
+            now = int(time.time() * 1000)
+            if cached['valid_until_ts'] > now:
+                defer.returnValue(self.cache[server_name]['verify_keys'])
+
+        host_port = yield self._getEndpointForServer(server_name)
+        logger.info("Got host/port %s/%s for %s", host_port[0], host_port[1], server_name)
+        client = FederationHttpClient(self.sydent)
+        result = yield client.get_json("https://%s:%s/_matrix/key/v2/server/" % host_port)
+        if 'verify_keys' not in result:
+            raise Exception("No key found in response")
+
+        if 'valid_until_ts' in result:
+            # Don't cache anything without a valid_until_ts or we wouldn't
+            # know when to expire it.
+            logger.info("Got keys for %s: caching until %s", server_name, result['valid_until_ts'])
+            self.cache[server_name] = result
+
+        defer.returnValue(result['verify_keys'])
+
+    @defer.inlineCallbacks
+    def verifyServerSignedJson(self, signed_json, acceptable_server_names=None):
+        """Given a signed json object, try to verify any one
+        of the signatures on it
+        XXX: This contains a very noddy version of the home server
+        SRV lookup and signature verification. It forms HTTPS URLs
+        from the result of the SRV lookup which will mean the Host:
+        parameter in the request will be wrong. It only looks at
+        the first SRV result. It does no caching (just fetches the
+        signature each time and does not contact any other servers
+        to do perspectives checks.
+
+        :param acceptable_server_names If provided and not None,
```

Comment:
of *course* we use different docstring conventions in different projects. Anyway there should be a colon after the param name aiui. And ideally documentation of a type.

https://pythonhosted.org/an_example_pypi_project/sphinx.html#function-definitions

Reply:
done.